find_exponential/src/main.rs

```diff
@@ -0,0 +1,270 @@
+use std::fs;
+use std::path::Path;
+use std::collections::HashSet;
+use std::time::{Duration, Instant};
+use std::panic;
+use std::sync::mpsc;
+use std::env;
+use std::io::{self, BufRead};
+
+use walkdir::WalkDir;
+use itertools::Itertools;
+use pulldown_cmark::{Parser, Options};
+use rand::seq::SliceRandom;
+use threadpool::ThreadPool;
+
+mod literals;
+mod black_box;
+mod scoring;
+
+/// How many combinations are batch-processed.
+const BATCH_SIZE: usize = 10_000;
+/// Combination length.
+const COMBINATIONS: usize = 4;
+/// If parsing takes longer than this many milliseconds, it is assumed to be non-linear.
+const MAX_MILLIS: u128 = 500;
+/// Size of repetition in bytes to parse.
+const NUM_BYTES: usize = 256*1024;
+/// Number of samples per pattern tested.
+const SAMPLE_SIZE: usize = 5;
+/// Score function to use when figuring out if something is non-linear.
+/// `scoring::{slope_stddev,pearson_correlation}`
+const SCORE_FUNCTION: fn(&Vec<[f64; 2]>) -> (f64, bool) = scoring::slope_stddev;
+/// If slope_stddev is used, if the standard deviation is larger than this, it's assumed to be non-linear.
+const ACCEPTANCE_STDDEV: f64 = 20.0;
+/// If pearson_correlation is used, if the correlation coefficient is below this value,
+/// it's assumed to be non-linear.
+const ACCEPTANCE_CORRELATION: f64 = 0.995;
+/// 0 / 1 / 2 / 3
+const DEBUG_LEVEL: u8 = 0;
+
+fn main() {
+    let num_cpus = num_cpus::get() * 3 / 4;
+
+    let arg = env::args().nth(1);
+    if let Some("--retest") = arg.as_ref().map(|s| s.as_str()) {
+        for pattern in io::stdin().lock().lines() {
+            let pattern = pattern.unwrap();
+            println!("retesting {:?}", pattern);
+            let score = test(&pattern);
+            println!("score: {}", score);
+        }
+        return;
+    } else {
+        fuzz(num_cpus);
+    }
+}
+
+fn fuzz(num_cpus: usize) {
+    // get all literals from pulldown-cmark's source code
+    let walkdir = WalkDir::new("../src")
+        .into_iter()
+        .filter_map(|e| e.ok())
+        .filter(|e| e.file_type().is_file())
+        .filter(|e| if let Some(ext) = e.path().extension() { ext == "rs" } else { false });
+
+    let mut literals = HashSet::new();
+    for file in walkdir {
+        if file.path() == Path::new("../src/entities.rs") {
+            continue;
+        }
+        println!("parsing {}", file.path().display());
+        let content = fs::read_to_string(file.path())
+            .expect(&format!("unable to read file {:?}", file.path()));
+        let parsed = syn::parse_file(&content)
+            .expect(&format!("unable to parse file {:?}", file.path()));
+        literals::extract_literals_from_items(&mut literals, parsed.items);
+    }
+
+    let literals: Vec<_> = literals.into_iter()
+        .filter(|lit| !lit.contains(&b'\n'))
+        .filter(|lit| !lit.is_empty())
+        .collect();
+
+    let pool = ThreadPool::new(num_cpus);
```

Comment:
No need for a thread-pool. I can just use crossbeam's mpmc channel and start `num_cpu` threads with a bounded channel of `num_cpu` elements.

Reply:
Actually mpmc / spmc isn't needed here. The only thing sent through it are the literals, which don't change between thread iterations. A simple `std::sync::mpsc::channel` is enough to send feedback from the worker thread back to the main thread. The main thread then print the "status update" with the throughput.

I could optimize it even further by making the count-variable atomic, incrementing it from each thread and printing the status update in the worker threads. Then I can just join on all threads in the main thread, letting it sleep forever (unless there are uncaught panics in itertools / std). This is probably a premature optimization though…

Reply:
I did the latter and used crossbeam scoped-threads. Done in 25dea8478cd43f6953b843523cf99abe288ed21e.